src/mlpack/methods/ann/layer/concatenate_impl.hpp

```diff
@@ -19,91 +19,106 @@
+ConcatenateType<MatType>&
+ConcatenateType<MatType>::operator=(const ConcatenateType& other)
 {
-  if (this != &layer)
+  if (&other != this)
   {
-    inRows = layer.inRows;
-    weights = layer.weights;
-    delta = layer.delta;
-    concat = layer.concat;
+    Layer<MatType>::operator=(other);
+    concat = other.concat;
   }
 
   return *this;
 }
 
-template<typename InputDataType, typename OutputDataType>
-Concatenate<InputDataType, OutputDataType>&
-Concatenate<InputDataType, OutputDataType>::
-operator=(Concatenate&& layer)
+template<typename MatType>
+ConcatenateType<MatType>&
+ConcatenateType<MatType>::operator=(ConcatenateType&& other)
 {
-  if (this != &layer)
+  if (&other != this)
   {
-    inRows = layer.inRows;
-    weights = std::move(layer.weights);
-    delta = std::move(layer.delta);
-    concat = std::move(layer.concat);
+    Layer<MatType>::operator=(std::move(other));
+    concat = std::move(other.concat);
   }
+
   return *this;
 }
 
-template<typename InputDataType, typename OutputDataType>
-template<typename eT>
-void Concatenate<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+template<typename MatType>
+void ConcatenateType<MatType>::Forward(const MatType& input, MatType& output)
 {
   if (concat.is_empty())
-    Log::Warn << "The concat matrix has not been provided." << std::endl;
-
-  if (input.n_cols != concat.n_cols)
   {
-    Log::Fatal << "The number of columns of the concat matrix should be equal "
-        << "to the number of columns of input matrix." << std::endl;
+    Log::Warn << "Concatenate::Forward(): the concat matrix is empty or was "
+        << "not provided." << std::endl;
   }
 
-  inRows = input.n_rows;
-  output = arma::join_cols(input, concat);
+  output.submat(0, 0, input.n_rows - 1, input.n_cols - 1) = input;
+  output.submat(input.n_rows, 0, output.n_rows - 1, input.n_cols - 1) =
+      arma::repmat(arma::vectorise(concat), 1, input.n_cols);
+}
+
+template<typename MatType>
+void ConcatenateType<MatType>::Backward(
+    const MatType& /* input */,
+    const MatType& gy,
+    MatType& g)
+{
+  // Pass back the non-concatenated part.
+  g = gy.submat(0, 0, gy.n_rows - 1 - concat.n_elem, gy.n_cols - 1);
 }
 
-template<typename InputDataType, typename OutputDataType>
-template<typename eT>
-void Concatenate<InputDataType, OutputDataType>::Backward(
-    const arma::Mat<eT>& /* input */,
-    const arma::Mat<eT>& gy,
-    arma::Mat<eT>& g)
+template<typename MatType>
+void ConcatenateType<MatType>::ComputeOutputDimensions()
 {
-  g = gy.submat(0, 0, inRows - 1, concat.n_cols - 1);
+  // This flattens the input.
+  size_t inSize = this->inputDimensions[0];
+  for (size_t i = 1; i < this->inputDimensions.size(); ++i)
+      inSize *= this->inputDimensions[i];
+
+  this->outputDimensions = std::vector<size_t>(this->inputDimensions.size(),
+      1);
+  this->outputDimensions[0] = inSize + concat.n_elem;
+}
+
+  /**
+   * Serialize the layer.
+   */
```

Comment:
```suggestion
/**
 * Serialize the layer.
 */
```

Or maybe remove it right away, I think most of the time we have the documentation in the header file only.